### PR TITLE
[WIP][Training] Flux Control LoRA training script

### PR DESCRIPTION
**Pull request by sayakpaul**
Will be updated later. 

This PR has both LoRA and a non-LoRA version for training Control Flux. 

~Number of changed files are more than expected because the `flux-control-lora` branch in merged into this PR branch.~

### CODE REVIEW

**Comment by a-r-r-o-w** [comment]
@sayakpaul I think some things like the model card are yet to be updated, and there's probably more things on your mind. Once you mark this ready for review, I can take a deeper look - everything looks great already though! Thanks for involving me in the discussions and looking forward to starting my own training run soon!

**Comment by sayakpaul** [comment]
> I think some things like the model card are yet to be updated

I think I have the model cards already updated. But let me know if you think some things should be added/updated. 

Marking this PR as ready to review.

**Comment by sayakpaul** [comment]
@apolinario @linoytsaban do you think it could make sense to show users how to _derive_ a LoRA from the full Control fine-tune like how the BFL folks did?

File: examples/control-lora/README.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# Training Control LoRA with Flux
+
+This (experimental) example shows how train Control LoRA with [Flux](https://huggingface.co/black-forest-labs/FLUX.1-dev) to condition it with additional structural controls (like depth maps, poses, etc.).
```
Comment (a-r-r-o-w) [inline examples/control-lora/README.md:3]:
```suggestion
This (experimental) example shows how to train Control LoRAs with [Flux](https://huggingface.co/black-forest-labs/FLUX.1-dev) by conditioning it with additional structural controls (like depth maps, poses, etc.).
```

File: examples/control-lora/README.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
+# Training Control LoRA with Flux
+
+This (experimental) example shows how train Control LoRA with [Flux](https://huggingface.co/black-forest-labs/FLUX.1-dev) to condition it with additional structural controls (like depth maps, poses, etc.). 
+
+We simply expand the input channels of Flux.1 Dev from 64 to 128 to allow for additional inputs and then train a regular LoRA on top of it. To account for the newly added input channels, we additional append a LoRA on the underlying layer (`x_embedder`). Inference, however, is performed with the `FluxControlPipeline`.
```
Comment (a-r-r-o-w) [inline examples/control-lora/README.md:5]:
```suggestion
To incorporate additional condition latents, we expand the input features of Flux.1-Dev from 64 to 128. The first 64 channels correspond to the original input latents to be denoised, while the latter 64 channels correspond to control latents. This expansion happens on the `x_embedder` layer, where the combined latents are projected to the expected feature dimension of rest of the network. Inference is performed using the `FluxControlPipeline`.
```

File: examples/control-lora/train_control_flux.py [hunk 3]
```diff
@@ -0,0 +1,1202 @@
...
+NORM_LAYER_PREFIXES = ["norm_q", "norm_k", "norm_added_q", "norm_added_k"]
+
+
+def encode_images(pixels: torch.Tensor, vae: torch.nn.Module, weight_dtype):
+    pixel_latents = vae.encode(pixels.to(vae.dtype)).latent_dist.sample()
+    pixel_latents = (pixel_latents - vae.config.shift_factor) * vae.config.scaling_factor
+    return pixel_latents.to(weight_dtype)
+
+
+def log_validation(flux_transformer, args, accelerator, weight_dtype, step, is_final_validation=False):
```
Comment (a-r-r-o-w) [inline examples/control-lora/train_control_flux.py:70]: Very clean!

File: examples/control-lora/train_control_flux.py [hunk 4]
```diff
@@ -0,0 +1,1202 @@
...
+            img_str += f"prompt: {validation_prompt}\n"
+            images = [validation_image] + images
+            make_image_grid(images, 1, len(images)).save(os.path.join(repo_folder, f"images_{i}.png"))
+            img_str += f"![images_{i})](./images_{i}.png)\n"
+
+    model_description = f"""
+# controlnet-{repo_id}
+
+These are Control weights trained on {base_model} with new type of conditioning.
+{img_str}
```
Comment (a-r-r-o-w) [inline examples/control-lora/train_control_flux.py:194]: Oh, this is what I missed earlier when I though model card was incomplete. The validation images will be displayed, awesome!

File: examples/control-lora/train_control_flux.py [hunk 5]
```diff
@@ -0,0 +1,1202 @@
...
+            validation_prompt = log["validation_prompt"]
+            validation_image = log["validation_image"]
+            validation_image.save(os.path.join(repo_folder, "image_control.png"))
+            img_str += f"prompt: {validation_prompt}\n"
+            images = [validation_image] + images
+            make_image_grid(images, 1, len(images)).save(os.path.join(repo_folder, f"images_{i}.png"))
+            img_str += f"![images_{i})](./images_{i}.png)\n"
+
+    model_description = f"""
+# controlnet-{repo_id}
```
Comment (a-r-r-o-w) [inline examples/control-lora/train_control_flux.py:191]:
```suggestion
# control-lora-{repo_id}
```

or something similar maybe

File: examples/control-lora/train_control_flux.py [hunk 6]
```diff
@@ -0,0 +1,1202 @@
...
+                    mode_scale=args.mode_scale,
+                )
+                indices = (u * noise_scheduler_copy.config.num_train_timesteps).long()
+                timesteps = noise_scheduler_copy.timesteps[indices].to(device=pixel_latents.device)
+
+                # Add noise according to flow matching.
+                sigmas = get_sigmas(timesteps, n_dim=pixel_latents.ndim, dtype=pixel_latents.dtype)
+                noisy_model_input = (1.0 - sigmas) * pixel_latents + sigmas * noise
+                # Concatenate across channels.
+                # Question: Should we concatenate before adding noise?
```
Comment (a-r-r-o-w) [inline examples/control-lora/train_control_flux.py:1032]: I think what you have currently is correct. Noisy latent + clean control latent

File: examples/control-lora/train_control_lora_flux.py [hunk 7]
```diff
@@ -0,0 +1,1353 @@
...
+        model_description=model_description,
+        inference=True,
+    )
+
+    tags = [
+        "flux",
+        "flux-diffusers",
+        "text-to-image",
+        "diffusers",
+        "control-lora",
```
Comment (a-r-r-o-w) [inline examples/control-lora/train_control_lora_flux.py:225]: I think also maybe add `lora` here to make it searchable on https://huggingface.co/models?other=lora as well. No problem if not

File: examples/control-lora/train_control_lora_flux.py [hunk 8]
```diff
@@ -0,0 +1,1353 @@
...
+        default=None,
+        choices=["no", "fp16", "bf16"],
+        help=(
+            "Whether to use mixed precision. Choose between fp16 and bf16 (bfloat16). Bf16 requires PyTorch >="
+            " 1.10.and an Nvidia Ampere GPU.  Default to the value of accelerate config of the current system or the"
+            " flag passed with the `accelerate.launch` command. Use this argument to override the accelerate config."
+        ),
+    )
+    parser.add_argument(
+        "--set_grads_to_none",
```
Comment (a-r-r-o-w) [inline examples/control-lora/train_control_lora_flux.py:446]: The default behaviour of pytorch is that grads will be set to None. With this, we default it to False unless specified. Is this intended? I think we should default to pytorch behaviour, and if an option is to be provided, it should be something like `dont_set_grads_to_none`. Not a blocker

File: examples/control-lora/train_control_lora_flux.py [hunk 9]
```diff
@@ -0,0 +1,1353 @@
...
+
+
+def main(args):
+    if args.report_to == "wandb" and args.hub_token is not None:
+        raise ValueError(
+            "You cannot use both --report_to=wandb and --hub_token due to a security risk of exposing your token."
+            " Please use `huggingface-cli login` to authenticate with the Hub."
+        )
+    if args.use_lora_bias and args.gaussian_init_lora:
+        raise ValueError("`gaussian` LoRA init scheme isn't supported when `use_lora_bias` is True.")
```
Comment (a-r-r-o-w) [inline examples/control-lora/train_control_lora_flux.py:729]: Just wondering why this is the case. We could set initialization to gaussian before the lora bias changes IIRC, no?

**Review by a-r-r-o-w — APPROVED** [review]
Awesome, LGTM! Just some minor comments

File: examples/control-lora/train_control_lora_flux.py [hunk 10]
```diff
@@ -0,0 +1,1353 @@
...
+        inference=True,
+    )
+
+    tags = [
+        "flux",
+        "flux-diffusers",
+        "text-to-image",
+        "diffusers",
+        "control-lora",
+        "diffusers-training",
```
Comment (sayakpaul) [inline examples/control-lora/train_control_lora_flux.py:226]:
```suggestion
        "diffusers-training",
        "lora",
```

File: examples/control-lora/train_control_lora_flux.py (hunk 8 above)
Comment (sayakpaul) [reply examples/control-lora/train_control_lora_flux.py:446]: You're right. I will remove it altogether.

File: examples/control-lora/train_control_lora_flux.py (hunk 9 above)
Comment (sayakpaul) [reply examples/control-lora/train_control_lora_flux.py:729]: So, we do:

```py
LoraConfig(lora_bias=True, init_lora_weights="gaussian", ...)
```

it errors out. Maybe a question for @BenjaminBossan.

File: examples/control-lora/train_control_lora_flux.py (hunk 7 above)
Comment (sayakpaul) [reply examples/control-lora/train_control_lora_flux.py]: Done.

**Comment by sayakpaul** [comment]
@a-r-r-o-w I have addressed your comments, thank you!

If you could review the changes and the README mods (made in https://github.com/huggingface/diffusers/pull/10130/commits/02016caf921e269cad4240b4b193f3831120bb65), that would be helpful!

**Review by a-r-r-o-w — APPROVED** [review]
Let's get the GPUs burning! LoRA testing scripts can be added in a follow-up PR if required but not very important

**Comment by sayakpaul** [comment]
Thanks, I think also okay to not add tests just yet. Based on the usage, we can always revisit.

**Comment by Adenialzz** [comment]
Hi, thanks for your implementation.

I'm curious why we set model offloading as default and don't have the option to turn it off. In my testing, this can significantly affect training speed ( ~5x). I recommend making model offloading optional in args

**Comment by sayakpaul** [comment]
@Adenialzz, thanks for your comment. Feel free to open a PR :)

**Comment by Adenialzz** [comment]
Sure. Here https://github.com/huggingface/diffusers/pull/10225 .

File: examples/control-lora/train_control_lora_flux.py (hunk 9 above)
Comment (BenjaminBossan) [reply examples/control-lora/train_control_lora_flux.py]: As [discussed earlier](https://github.com/huggingface/peft/pull/2237#discussion_r1860357307), it just hasn't been implemented yet. The question is: If a user wants Gaussian init for the LoRA weights, how do we initialize the LoRA bias? If we have an answer to that, we can enable Gaussian init for `lora_bias`.

**Pull request merged**